File: app/src/main/java/com/krystianwsul/checkme/domainmodel/TaskHierarchy.java
package com.krystianwsul.checkme.domainmodel;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.utils.TaskHierarchyKey;
import com.krystianwsul.checkme.utils.TaskKey;
import com.krystianwsul.checkme.utils.time.ExactTimeStamp;

public abstract class TaskHierarchy {
    @NonNull
    protected final DomainFactory mDomainFactory;

    protected TaskHierarchy(@NonNull DomainFactory domainFactory) {
        mDomainFactory = domainFactory;
    }

    public boolean current(@NonNull ExactTimeStamp exactTimeStamp) {
        ExactTimeStamp startExactTimeStamp = getStartExactTimeStamp();
        ExactTimeStamp endExactTimeStamp = getEndExactTimeStamp();

        return (startExactTimeStamp.compareTo(exactTimeStamp) <= 0 && (endExactTimeStamp == null || endExactTimeStamp.compareTo(exactTimeStamp) > 0));
    }

    @NonNull
    protected abstract ExactTimeStamp getStartExactTimeStamp();

    @Nullable
    protected abstract ExactTimeStamp getEndExactTimeStamp();

    public abstract double getOrdinal();

    @NonNull
    public abstract TaskKey getParentTaskKey();

    @NonNull
    public abstract TaskKey getChildTaskKey();

    @NonNull
    public abstract Task getParentTask();

    @NonNull
    public abstract Task getChildTask();

    boolean notDeleted(@NonNull ExactTimeStamp exactTimeStamp) {
        ExactTimeStamp endExactTimeStamp = getEndExactTimeStamp();

        return (endExactTimeStamp == null || endExactTimeStamp.compareTo(exactTimeStamp) > 0);
    }

    public abstract void setEndExactTimeStamp(@NonNull ExactTimeStamp now);

    public abstract void setOrdinal(double ordinal);

    public abstract void delete();

    @NonNull
    public abstract TaskHierarchyKey getTaskHierarchyKey();
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/TaskHierarchy.kt
package com.krystianwsul.checkme.domainmodel

import com.krystianwsul.checkme.utils.TaskHierarchyKey
import com.krystianwsul.checkme.utils.TaskKey
import com.krystianwsul.checkme.utils.time.ExactTimeStamp

abstract class TaskHierarchy protected constructor(protected val mDomainFactory: DomainFactory) {

    abstract val startExactTimeStamp: ExactTimeStamp

    protected abstract fun getEndExactTimeStamp(): ExactTimeStamp?

    abstract var ordinal: Double

    abstract val parentTaskKey: TaskKey

    abstract val childTaskKey: TaskKey

    abstract val parentTask: Task

    abstract val childTask: Task

    abstract val taskHierarchyKey: TaskHierarchyKey

    fun current(exactTimeStamp: ExactTimeStamp): Boolean {
        val startExactTimeStamp = startExactTimeStamp
        val endExactTimeStamp = getEndExactTimeStamp()

        return startExactTimeStamp <= exactTimeStamp && (endExactTimeStamp == null || endExactTimeStamp > exactTimeStamp)
    }

    fun notDeleted(exactTimeStamp: ExactTimeStamp): Boolean {
        val endExactTimeStamp = getEndExactTimeStamp()

        return endExactTimeStamp == null || endExactTimeStamp > exactTimeStamp
    }

    abstract fun setEndExactTimeStamp(now: ExactTimeStamp)

    abstract fun delete()
}
